pkg/controller/core/admissioncheck_controller.go

```diff
@@ -0,0 +1,230 @@
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(ac, kueue.ResourceInUseFinalizerName) {
+			if cqs := r.cache.ClusterQueuesUsingAdmissionCheck(ac.Name); len(cqs) != 0 {
+				log.V(3).Info("admissionCheck is still in use", "ClusterQueues", cqs)
+				// We avoid to return error here to prevent backoff requeue, which is passive and wasteful.
+				// Instead, we drive the removal of finalizer by ClusterQueue Update/Delete events
+				// when the admissionCheck is no longer in use.
+				return ctrl.Result{}, nil
+			}
+			controllerutil.RemoveFinalizer(ac, kueue.ResourceInUseFinalizerName)
+			if err := r.client.Update(ctx, ac); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.V(5).Info("Removed finalizer")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+func (r *AdmissionCheckReconciler) notifyWatchers(oldAc, newAc *kueue.AdmissionCheck) {
+	for _, w := range r.watchers {
+		w.NotifyAdmissionCheckUpdate(oldAc, newAc)
+	}
+}
+
+func (r *AdmissionCheckReconciler) AddUpdateWatchers(watchers ...AdmissionCheckUpdateWatcher) {
+	r.watchers = append(r.watchers, watchers...)
+}
+
+func (r *AdmissionCheckReconciler) Create(e event.CreateEvent) bool {
+	ac, isAc := e.Object.(*kueue.AdmissionCheck)
+	if !isAc {
+		return false
+	}
+	defer r.notifyWatchers(nil, ac)
+	r.log.WithValues("admissionCheck", klog.KObj(ac)).V(5).Info("Create event")
+	if cqNames := r.cache.AddOrUpdateAdmissionCheck(ac); len(cqNames) > 0 {
+		r.qManager.QueueInadmissibleWorkloads(context.Background(), cqNames)
+	}
+	return true
+}
+
+func (r *AdmissionCheckReconciler) Update(e event.UpdateEvent) bool {
+	newAc, isAc := e.ObjectNew.(*kueue.AdmissionCheck)
+	if !isAc {
+		return false
+	}
+	oldAc, isAc := e.ObjectNew.(*kueue.AdmissionCheck)
+	if !isAc {
+		return false
+	}
+	defer r.notifyWatchers(oldAc, newAc)
+	r.log.WithValues("admissionCheck", klog.KObj(newAc)).V(5).Info("Update event")
+	if !newAc.DeletionTimestamp.IsZero() {
+		return true
+	}
+	if cqNames := r.cache.AddOrUpdateAdmissionCheck(newAc); len(cqNames) > 0 {
+		r.qManager.QueueInadmissibleWorkloads(context.Background(), cqNames)
+	}
+	return false
+}
+
+func (r *AdmissionCheckReconciler) Delete(e event.DeleteEvent) bool {
+	ac, isAc := e.Object.(*kueue.AdmissionCheck)
+
+	if !isAc {
+		return false
+	}
+	defer r.notifyWatchers(ac, nil)
+	r.log.WithValues("admissionCheck", klog.KObj(ac)).V(5).Info("Delete event")
+
+	if cqNames := r.cache.DeleteAdmissionCheck(ac); len(cqNames) > 0 {
+		r.qManager.QueueInadmissibleWorkloads(context.Background(), cqNames)
+	}
+	return true
+}
+
+func (r *AdmissionCheckReconciler) Generic(e event.GenericEvent) bool {
+	r.log.WithValues("object", klog.KObj(e.Object), "kind", e.Object.GetObjectKind().GroupVersionKind()).V(5).Info("Generic event")
+	return true
+}
+
+func (r *AdmissionCheckReconciler) NotifyClusterQueueUpdate(oldCq *kueue.ClusterQueue, newCq *kueue.ClusterQueue) {
+	log := r.log.WithValues("oldCq", klog.KObj(oldCq), "newCq", klog.KObj(newCq))
```

Comment:
```suggestion
	log := r.log.WithValues("oldClusterQueue", klog.KObj(oldCq), "newClusterQueue", klog.KObj(newCq))
```

We shouldn't drop the context in logs. Let's keep using `clusterQueue` the same as other locations.